File: sdk/ruby/test/wire/labels_test.rb
# frozen_string_literal: true

require_relative "wiremock_test_case"

class LabelsWireTest < WireMockTestCase
  def setup
    super

    @client = AgentMailer::Client.new(
      api_key: "<token>",
      base_url: WIREMOCK_BASE_URL
    )
  end

  def test_labels_list_with_wiremock
    test_id = "labels.list.0"

    @client.labels.list(request_options: {
      additional_headers: {
        "X-Test-Id" => "labels.list.0"
      }
    })

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/labels",
      query_params: nil,
      expected: 1
    )
  end

  def test_labels_create_with_wiremock
    test_id = "labels.create.0"

    @client.labels.create(
      pod_id: "podId",
      name: "name",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "labels.create.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/labels",
      query_params: nil,
      expected: 1
    )
  end

  def test_labels_delete_with_wiremock
    test_id = "labels.delete.0"

    @client.labels.delete(
      label_id: "labelId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "labels.delete.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "DELETE",
      url_path: "/v1/labels/labelId",
      query_params: nil,
      expected: 1
    )
  end

  def test_labels_update_message_with_wiremock
    test_id = "labels.update_message.0"

    @client.labels.update_message(
      message_id: "messageId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "labels.update_message.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "PATCH",
      url_path: "/v1/messages/messageId/labels",
      query_params: nil,
      expected: 1
    )
  end

  def test_labels_update_thread_with_wiremock
    test_id = "labels.update_thread.0"

    @client.labels.update_thread(
      thread_id: "threadId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "labels.update_thread.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "PATCH",
      url_path: "/v1/threads/threadId/labels",
      query_params: nil,
      expected: 1
    )
  end
end

File: sdk/rust/tests/labels_test.rs
use agentmailer::prelude::*;

mod wire_test_utils;

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_labels_list_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .labels
        .list(
            &LabelsListQueryRequest {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("GET", "/v1/labels", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_labels_create_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .labels
        .create(
            &LabelCreate {
                pod_id: "podId".to_string(),
                name: "name".to_string(),
                color: None,
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/labels", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_labels_delete_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client.labels.delete(&"labelId".to_string(), None).await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("DELETE", "/v1/labels/labelId", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_labels_update_message_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .labels
        .update_message(
            &"messageId".to_string(),
            &LabelMutation {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("PATCH", "/v1/messages/messageId/labels", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_labels_update_thread_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .labels
        .update_thread(
            &"threadId".to_string(),
            &LabelMutation {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("PATCH", "/v1/threads/threadId/labels", None, 1)
        .await
        .unwrap();
}
